pegasus/sites.v3/hourofcode.com/views/signup_form.haml

```diff
@@ -6,109 +7,280 @@
+        %label.control-label{for: "school-name"}
+          = hoc_s(:school_name)
+          %span.required*
+        %div
+          %input#school-name.form-control{name: "school_name_s", type: "text"}
+        #school-name-error.error-message{style: "display:none;"}
+          = hoc_s(:school_name_error)
+
+      .form-group#organization-name-field{style: "display: none;"}
+        %label.control-label{for: "hoc-organization-name"}
+          = hoc_s(:signup_organization_name_label)
+          %span.required*
+        %div
+          %input#organization-name.form-control{name: "organization_name_s", type: "text"}
+        #organization-name-error.error-message{style: "display:none;"}
+          = hoc_s(:signup_organization_name_error)
+
+      .form-group#hoc-event-location-field{style: "display: none;"}
+        %label.control-label{for: "hoc-event-location"}
+          = hoc_s(:signup_event_location_label)
+          %span.required*
+        %div
+          %textarea#hoc-event-location.form-control{name: "event_location_s", placeholder: "Event location address", style: "height: 4em;"}
+        #event-location-error.error-message{style: "display:none;"}
+          = hoc_s(:signup_event_location_error)
+
+      .form-group#hoc-entire-school{style: "display: none;"}
+        .checkbox
+          %label
+            %input#hoc-entire-school-flag{name: "entire_school_flag_b", type: "checkbox", value: "1"}/
+            = hoc_s(:signup_event_entire_school_label)
+
+      .form-group
+        .checkbox
+          %label
+            %input#hoc-special-event-flag{name: "special_event_flag_b", type: "checkbox", value: "1"}/
+            = hoc_s(:signup_special_event_flag_label)
+
+      .form-group{style: "display: none;"}
+        %label.control-label{for: "hoc-special-event-details"}
+          = hoc_s(:signup_special_event_details_label)
+        %div
+          %textarea#hoc-special-event-details.form-control{name: "special_event_details_s", placeholder: hoc_s(:signup_special_event_details_placeholder), style: "height: 5em;"}
+
+      .form-group
+        .checkbox
+          %label
+            %input#hoc-match-volunteer-flag{name: "match_volunteer_flag_b", type: "checkbox", value: "1"}/
+            = hoc_s(:signup_match_volunteer)
+
+    .form-group.continue-btn{style: "display: none;"}
       %div
-        %button{type: "submit"}= hoc_s(:signup_submit_label)
+        %button.continue-btn{type: "button"}
+          = hoc_s(:signup_continue_label)
+
+    #census-questions{style: "display: none;"}
+      =view :census_questions
+
+    #submit
+      %input{name:'hoc_country_s', type:'hidden', value:@country}
+      %input{name:'hoc_company_s', type:'hidden', value:@company}
+      %p.footnote
+        = hoc_s(:signup_multiple_event_warning)
+        %strong= hoc_s(:signup_registration_deadline)
+      .form-group.submit-btn
+        %div
+          %button{type: "submit"}= hoc_s(:signup_submit_label)
 
 :javascript
```

Comment:
I'm aware that it would be cleaner to move this JavaScript to a separate file that could be run through a linter and tested; however some of these JavaScript functions rely on Ruby within this haml file. For example, the ajax call in `signupFormSubmit()` needs to know what `hoc_year` is, the thank you redirect relies on resolving a url, and `signupFormError(data)` uses Ruby as well.  Then it cascades - for instance, the submit function is dependent on the validation function. Thus is seemed even worse to have some JavaScript in the haml file and some living in a separate file, so I left it here.

Reply:
The way we've handled this in the past is to have just enough javascript in the ruby code to define javascript variables to hold the values of the ruby variables. Your code here would look like this:
```
:javascript
  var thanksUrl = "#{resolve_url('thanks')}";
  var signupErrorMessage = "#{signup_submit_error_message}";
```
Then these would be accessible as global variables from within the JS file you extract. You don't really have to worry about it "cascading" -- in either approach, the ruby is being executed when the page is rendered, and inserting the result into the javascript code as flat strings.

I'd be happy to answer any questions on slack if you'd like to take another crack at getting this working. However, I know you're wading through a lot of technical debt that you didn't create, so I think it's ok to punt on this if you feel like you're spending too much of your time refactoring compared to the amount of time spent on the actual feature development.

Reply:
to add a bit more detail, your definition of `signupFormError` would then look like:
```javascript
  function signupFormError(data) {
    $('#error_message').html("<p>" + signupErrorMessage + "</p>").show();
  ...
```

Reply:
Thanks for the suggestion. This approach is way better! I think I got it working.